File: C#/第二套/11900-1060306/11900-1060306/Presenters/MainPresenter.cs
namespace IdCardChecker.Presenters;

using IdCardChecker.Views;
using IdCardChecker.Services;
using IdCardChecker.Models;

public class MainPresenter
{
    private readonly IMainView _view;
    private readonly IIdCardValidatorService _validator;
    private readonly IDataService _dataService;

    public MainPresenter(IMainView view, IIdCardValidatorService validator, IDataService dataService)
    {
        _view = view;
        _validator = validator;
        _dataService = dataService;

        _view.LoadDataClicked += OnLoadDataClicked;
    }

    private async void OnLoadDataClicked(object? sender, EventArgs e)
    {
        using OpenFileDialog openFileDialog = new()
        {
            Filter = "SM/T01 Files (*.SM;*.T01)|*.SM;*.T01|All Files (*.*)|*.*",
            Title = "Select ID Card Data File"
        };

        if (openFileDialog.ShowDialog((IWin32Window)_view) == DialogResult.OK)
        {
            var records = await _dataService.ReadRecordsAsync(openFileDialog.FileName);
            
            var validatedRecords = records
                .Select(r => _validator.Validate(r))
                .OrderBy(r => r.Id)
                .ToList();

            _view.ShowRecords(validatedRecords);
        }
    }
}

File: VB/第二套/11900-1060306/Presenters/MainPresenter.vb
Namespace Presenters
    Public Class MainPresenter
        Private ReadOnly _view As Views.IMainView
        Private ReadOnly _validator As Services.IIdCardValidatorService
        Private ReadOnly _dataService As Services.IDataService

        Public Sub New(view As Views.IMainView, validator As Services.IIdCardValidatorService, dataService As Services.IDataService)
            _view = view
            _validator = validator
            _dataService = dataService
            
            ' Initialize View Data
            _view.CandidateName = "陳宇威"
            _view.CandidateNumber = "112590005"
            _view.DeskNumber = "005"
            _view.TestDate = $"{DateTime.Now:yyyy/MM/dd}"

            
            AddHandler _view.LoadDataRequested, AddressOf OnLoadDataRequested
        End Sub

        Private Sub OnLoadDataRequested(sender As Object, filePath As String)
            Dim records = _dataService.LoadData(filePath)
            Dim results = New List(Of Models.IdCardRecord)()
            For Each r In records
                results.Add(_validator.Validate(r))
            Next
            
            ' Sort by ID_NO (lexicographical order)
            Dim sortedResults = results.OrderBy(Function(r) r.Id).ToList()
            _view.SetResults(sortedResults)
        End Sub

    End Class
End Namespace
